File: 133-clone-graph/133-clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

# edge. The original node could be None
# 1. Put the first node in the queue
# 2. Use a dictionary to save the visited node and it's respective clone as key and value => avoid cycles
# 3. Clone the node and put it in the visited dictionary
# 4. Start BFS traversal
    # pop a node cur from the front of the queue
    # iterate through all the neighbors of the cur node
      # clone the nei and put it as visited if not present already
      # add the newly encountered node to the queue
    # add the clone of the nei to the neighbors of the clone node q
# 5. Return the clone of the node from visited 
    
    
class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if not node: return node 
        
        q = deque([node])
        visited = {}
        visited[node] = Node(node.val, [])
        
        #BFS
        while q:
            cur = q.popleft()
            
            for nei in cur.neighbors:
                if nei not in visited:
                    visited[nei] = Node(nei.val, [])
                    q.append(nei)
                visited[cur].neighbors.append(visited[nei])   
                
        return visited[node]
        

File: 133-clone-graph/133-clone-graph.js
// Approach: Breadth First Search
// Time complexity: O(N)
// Space complexity: O(N)


/**
 * // Definition for a Node.
 * function Node(val, neighbors) {
 *    this.val = val === undefined ? 0 : val;
 *    this.neighbors = neighbors === undefined ? [] : neighbors;
 * };
 */

/**
 * @param {Node} node
 * @return {Node}
 */
var cloneGraph = function(node) {
    if (node === null) {
        return null;
    }
    
    let visited = new Map();
    let q = [node];
    visited.set(node, new Node(node.val, [])) ;
    
    // BFS
    while (q.length > 0) {
        let cur = q.shift();
        for (let nei of cur.neighbors) {
            if (!visited.has(nei)) {
                visited.set(nei, new Node(nei.val, []));
                q.push(nei);
            }
            visited.get(cur).neighbors.push(visited.get(nei));
        } 
    }
    return visited.get(node);
};